src/client.cpp

```diff
@@ -291,31 +281,90 @@ void CClient::OnNewConnection()
+    // ALSO here is where we allocate and free client channels as required
 
-    // Reset all gains for unused channel IDs:
-    for ( int iId = 0; iId < MAX_NUM_CHANNELS; iId++ )
+    const int iNumConnectedClients = vecChanInfo.Size();
+    int       i, iSrvIdx;
+
+    // on a new connection, a server sends an empty channel list before sending
+    // the real channel list (see #1010). To avoid this discarding "our" channel
+    // that we have just created, we skip this processing and just pass the empty
+    // list to the emitted signal.
+
+    if ( iNumConnectedClients != 0 )
     {
-        if ( !bChanIdInUse[iId] )
+        // this relies on the received client list being in order of server channel ID
+
+        for ( i = 0, iSrvIdx = 0; i < iNumConnectedClients && iSrvIdx < MAX_NUM_CHANNELS; )
         {
-            // reset oldGain and newGain as this channel id is currently unused and will
-            // start with a server-side gain at 1 (100%) again.
-            oldGain[iId] = newGain[iId] = 1;
+            // server channel ID of this entry
+            const int iServerChannelID = vecChanInfo[i].iChanID;
+
+            // find matching client channel ID, creating new if necessary
+            const int iClientChannelID = FindClientChannel ( iServerChannelID, true );
+
+            // discard any lower server channels that are no longer in our local list
+            while ( iSrvIdx < iServerChannelID )
+            {
+                const int iId = FindClientChannel ( iSrvIdx, false );
+
+                if ( iId != INVALID_INDEX )
+                {
+                    // iSrvIdx contains a server channel number that has now gone
+                    FreeClientChannel ( iSrvIdx );
+                }
+                iSrvIdx++;
+            }
+
+            // now should have matching server channel IDs
+            vecChanInfo[i].iChanID = iClientChannelID; // update channel number to be client-side
+            i++;                                       // next list entry
+            iSrvIdx++;                                 // next local server channel
         }
+
+        // have now run out of active channels, discard any remaining from our local list
+        // note that iActiveChannels will reduce as remaining channels are freed
+
+        while ( iActiveChannels > iNumConnectedClients && iSrvIdx < MAX_NUM_CHANNELS )
+        {
+            const int iId = FindClientChannel ( iSrvIdx, false );
+
+            if ( iId != INVALID_INDEX )
+            {
+                // iSrvIdx contains a server channel number that has now gone
+                FreeClientChannel ( iSrvIdx );
+            }
+            iSrvIdx++;
+        }
+
+        Q_ASSERT ( iActiveChannels == iNumConnectedClients );
     }
+
+    // TODO Does vecChanInfo need to be ordered by client channel ID instead of server channel ID?
+    // check if this is actually necessary, and whether it affects the level meters
```

Comment:
It doesn't actually appear to be necessary. The updating of channel IDs in-place seems to be sufficient. But I'd be interested in others' opinions.

Reply:
So long as the "No fader sort" continues to add newly joining clients to the right-hand end of the mixer GUI, it's working okay.  If a newly joining client jolts anyone already on rightwards and slots in to the left, that's not okay.

Reply:
Yes, I've just done some testing with "No fader sort". I added several clients, and then disconnected some to leave gaps in the used channel numbers. Re-adding them used the channel numbers that were vacated, but the faders were definitely added on the right in order of arrival.

Reply:
In view of that, I'll remove the above comment from the code.